// rrouter/lib/routing/DummyRouting.js
/**
 * @jsx React.DOM
 */
'use strict';

var Routing = require('./Routing');

for(var Routing____Key in Routing){if(Routing.hasOwnProperty(Routing____Key)){DummyRouting[Routing____Key]=Routing[Routing____Key];}}var ____SuperProtoOfRouting=Routing===null?null:Routing.prototype;DummyRouting.prototype=Object.create(____SuperProtoOfRouting);DummyRouting.prototype.constructor=DummyRouting;DummyRouting.__superConstructor__=Routing;

  function DummyRouting(routes, match, path, query) {
    this.routes = routes;
    this.match = match;
    this.path = path;
    this.query = query;
  }

  DummyRouting.prototype.getPath=function() {
    return this.path;
  };

  DummyRouting.prototype.getQuery=function() {
    return this.query;
  };

  DummyRouting.prototype.update=function() {
    throw new Error('not implemented');
  };

  DummyRouting.prototype.pushPath=function() {
    throw new Error('not implemented');
  };

  DummyRouting.prototype.replacePath=function() {
    throw new Error('not implemented');
  };

  DummyRouting.prototype.doStart=function() {
    throw new Error('not implemented');
  };

  DummyRouting.prototype.doStop=function() {
    throw new Error('not implemented');
  };


module.exports = DummyRouting;
