// rrouter/lib/routing/HashRouting.js
/**
 * @jsx React.DOM
 */
'use strict';

var Routing = require('./Routing');

for(var Routing____Key in Routing){if(Routing.hasOwnProperty(Routing____Key)){HashRouting[Routing____Key]=Routing[Routing____Key];}}var ____SuperProtoOfRouting=Routing===null?null:Routing.prototype;HashRouting.prototype=Object.create(____SuperProtoOfRouting);HashRouting.prototype.constructor=HashRouting;HashRouting.__superConstructor__=Routing;function HashRouting(){if(Routing!==null){Routing.apply(this,arguments);}}

  HashRouting.prototype.getPath=function() {
    return this.getParsedPath().path;
  };

  HashRouting.prototype.getQuery=function() {
    return this.getParsedPath().query;
  };

  HashRouting.prototype.getParsedPath=function() {
    var path = window.location.hash.slice(1) || '/';
    var idx = path.indexOf('?');
    if (idx > -1) {
      return {path: path.substring(0, idx), query: path.substring(idx + 1)};
    } else {
      return {path:path, query: ''};
    }
  };

  HashRouting.prototype.pushPath=function(path) {
    window.location.hash = path;
  };

  HashRouting.prototype.replacePath=function(path) {
    var href = window.location.href.replace(/(javascript:|#).*$/, '');
    window.location.replace(href + '#' + path);
  };

  HashRouting.prototype.makeHref=function(name, params) {
    return '#' + ____SuperProtoOfRouting.makeHref.call(this,name, params);
  };

  HashRouting.prototype.doStart=function() {
    window.addEventListener('hashchange', this.onBackButton);
  };

  HashRouting.prototype.doStop=function() {
    window.removeEventListener('hashchange', this.onBackButton);
  };


module.exports = HashRouting;
