// rrouter/lib/routing/Routing.js
/**
 * @jsx React.DOM
 */
'use strict';

var qs                    = require('qs');
var React                 = require('react');
var createView            = require('../createView');
var matchRoutes           = require('../matchRoutes');
var data                  = require('../data');
var fetchViews            = require('../fetchViews');
var makeHref              = require('../makeHref');

function throwError(err) {
  throw err;
}



  function Routing(routes, onRoute, onError) {
    this.routes = routes;
    this.onRoute = onRoute;
    this.onError = onError || throwError;
    this.onChange = this.onChange.bind(this);
    this.onBackButton = this.onBackButton.bind(this);
    this.path = undefined;
    this.match = undefined;
    this.started = false;
  }

  Routing.prototype.onChange=function(navigation) {
    navigation = navigation || {};
    var path = this.getPath();
    var query = this.getQuery();
    if (this.path !== path || this.query !== query) {
      this.$Routing_route(path, query, navigation);
    }
  };

  Routing.prototype.onBackButton=function() {
    this.onChange({back: true});
  };

  Routing.prototype.update=function() {
    this.$Routing_route(this.getPath(), this.getQuery());
  };

  Routing.prototype.$Routing_route=function(path, query, navigation) {
    this.path = path;
    this.query = query;
    this.match = matchRoutes(this.routes, path, query);

    var expectedMatch = this.match;

    var render = function(match)  {
      if (this.match === expectedMatch) {
        this.renderView(match, navigation);
      }
    }.bind(this);

    var promise = fetchViews(this.match);

    if (promise.isFulfilled()) {
      render(data.fetchProgressively(promise.value(), render, this.onError));
    } else {
      promise.then(function(match)  {
        render(data.fetchProgressively(match, render, this.onError));
      }.bind(this), this.onError);
    }

    return this.match;
  };

  Routing.prototype.renderView=function(match, navigation) {
    var context = {match:match, routing: this, routes: this.routes};
    React.withContext(context, function()  {
      var view = createView(match);
      this.onRoute(view, match, navigation);
    }.bind(this));
  };

  Routing.prototype.navigate=function(path, navigation) {
    navigation = navigation || {};
    if (navigation.replace) {
      this.replacePath(path, navigation);
    } else {
      this.pushPath(path, navigation);
    }
    return this.onChange(navigation);
  };

  Routing.prototype.navigateTo=function(routeRef, props, navigation) {
    var path = this.makeHref(routeRef, props);
    this.navigate(path, navigation);
  };

  Routing.prototype.makeHref=function(name, params) {
    var href = makeHref(this.routes, name, this.match, params);
    if (params && params.query) {
      href = href + '?' + qs.stringify(params.query);
    }
    return href;
  };

  Routing.prototype.start=function() {
    if (!this.started) {
      this.doStart();
      this.onChange({initial: true});
      this.started = true;
    }
    return this;
  };

  Routing.prototype.stop=function() {
    if (this.started) {
      this.doStop();
      this.started = false;
    }
    return this;
  };

  Routing.start=function(routes, onRoute, onError) {
    return new this(routes, onRoute, onError).start();
  };


module.exports = Routing;
